docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,272 @@
+`MyFFIType` struct, so that layout can't be communicated to the foreign bindings.
+As noted above, this is considered a feature for diplomat, but a limitation for UniFFI.
+
+This is the problem which caused us to decide to stop working on
+[#416](https://github.com/mozilla/uniffi-rs/pull/416) - the current world where the type universe
+is described externally doesn't have this limitation - only the UDL file needs to be parsed when
+generating the foreign bindings. The application-services team has
+concluded that none of our non-trival use-cases for UniFFI could be described using macros,
+so supporting both mechanisms is pain for no gain.
+
+As noted in #416, `wasm-bindgen` has a similarly shaped problem, and solves it by having
+the Rust macro arrange for the resulting library to have an extra data section with the
+serialized "type universe" - foreign binding generation would then read this information from the
+already built binary. This sounds more complex than the UniFFI team has appetite for at
+the current time.
+
+# Looking forward
+
+## Adapting the diplomat/[#416](https://github.com/mozilla/uniffi-rs/pull/416) model to process the entire crate?
+
+We noted that diplomat intentionally restricts where the ffi is generated,
+whereas UniFFI considers that a limitation - but what if we can teach UniFFI to process
+more of the Rust crate?
+
+It might be reasonable for the foreign bindings to know that Rust "paths" to modules which should
+be processed, and inside those modules find structs "marked up" as being used by the FFI.
+
+In other words, borrowing the example above:
+
+```Rust
+#[uniffi::magic]
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+```
+
+maybe can made to work, so long as we are happy to help UniFFI discover where such annotations
+may exist.
+
+A complication here is that currently UniFFI allows types defined in external crates,
+but that might still be workable - eg,
+[diplomat has an issue open to support exactly this](https://github.com/rust-diplomat/diplomat/issues/34)
+
+## Duplicating structs inside Rust
+
+When reviewing the draft of this document, @rfk noted that we are already duplicating Rust structs
+in UDL and in Rust. So instead of having:
+
+```
+// In a UDL file:
+dictionary MyFFIType {
+    i32 a;
+    bool b;
+};
+// Then in Rust:
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+```
+
+we could have:
+```rust
+// In the Rust implementation, in some other module.
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+
+// And to expose it over the FFI:
+#[ffi::something]
+mod ffi {
+
+    #[ffi::magic_external_type_declaration]
+    pub struct MyFFIType {
+        pub a: i32,
+        pub b: bool,
+    }
+
+    impl MyFFIType {
+        pub fn create() -> MyFFIType { ... }
+        ...
+    }
+}
+```
+
+So while we haven't exactly reduced the duplication, we have removed the UDL.
+We probably also haven't helped with documentation, because the natural location for
```

Comment:
You could imagine explicitly telling the macro about the path to the redeclared struct, like:

```
    #[uniffi::declare_imported_type(super::MyFFIType)]
    pub struct MyFFIType {
        pub a: i32,
        pub b: bool,
    }
```

Then if it wanted to, the code processing this declaration could go find the corresponding `super::myFFIType` and pull out e.g. its docstring.

I think this would *probably* be more trouble than it's worth (we don't want to re-implement vast swathes of Rust's name lookup machinery, for example) but it's interesting to think about.